apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriptionHandler.cs

```diff
@@ -0,0 +1,78 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License").
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at 
+//
+// https://www.apache.org/licenses/LICENSE-2.0 
+//
+// Unless required by applicable law or agreed to in writing, software 
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and 
+// limitations under the License.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using static Google.Cloud.PubSub.V1.SubscriberClient;
+
+namespace Google.Cloud.PubSub.V1;
+
+/// <summary>
+/// This abstract class provides the message handler to handle and process the messages.
+/// It also provides the ACKnowledge (ACK) and Not ACKnowledge (NACK) response handlers which indicates the ACK/NACK status of each message for the subscription.
+/// The users of subscription should inherit from this class and provide their implementation to handle messages and their acknowledgement response.
+/// </summary>
+public abstract class SubscriptionHandler
+{
+    /// <summary>
+    /// The handler function that is passed all received messages.
+    /// This function may be called on multiple threads concurrently. Return <see cref="Reply.Ack"/> from this function
+    /// to ACKnowledge this message (implying it won't be received again); or return <see cref="Reply.Nack"/> to Not
+    /// ACKnowledge this message (implying it will be received again). If this function throws any Exception, then
+    /// it behaves as if it returned <see cref="Reply.Nack"/>.
+    /// </summary>
+    /// <remarks>
+    /// The users of subscription must implement this function to handle the messages.
+    /// </remarks>
+    public abstract Task<Reply> HandleMessage(PubsubMessage message, CancellationToken cancellationToken);
+
+    /// <summary>
+    /// This handler function is called when an ACKnowledge (ACK) operation returns, indicating the status of successful or permanently failed messages.
+    /// </summary>
+    /// <param name="responses">The read-only collection of the <see cref="AckNackResponse"/>
+    /// indicating the <see cref="SubscriberClient.Reply.Ack"/> status of successful or permanently failed messages. This collection may be empty but never null.
+    /// For exactly once delivery subscription, this collection contains only the successful or permanently failed messages. In this case,
```

Comment:
We should strongly suggest the user implement and handle ack/nack failures for EOS subs. The EOS guarantees don't hold if they don't deal with permanent ack/nack failures.

Reply:
Which remarks section? I don't see a change for the comments for this method, nor for the class comment.

Reply:
Please see line number 50-55. It has a remarks tag and documentation. Please let me know if any additional message needs to be added and I can do it.

Reply:
We need to indicate that when EOS is turned on, users SHOULD to implement that method. Otherwise it is optional.

Reply:
Updated the documentation.